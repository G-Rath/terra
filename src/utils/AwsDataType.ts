export enum AwsDataType {
  AWS_IAM_ROLE_POLICY_DOCUMENT = 'aws_iam_policy_document'
}
