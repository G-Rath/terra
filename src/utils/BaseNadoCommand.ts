import { Command, flags } from '@oclif/command';
import { format } from '@src/formatter';
import { printTFFileContents } from '@src/printer';
import {
  DataType,
  TFBlock,
  TFBlocks,
  TFFileContents,
  TFResourceBlock
} from '@src/types';
import dedent from 'dedent';
import { promises as fs } from 'fs';

export interface ParseResults {
  args: Record<string, unknown>;
  flags: {
    format?: boolean;
    write?: boolean;
    filepath?: string;
  };
}

export abstract class BaseNadoCommand extends Command {
  public static flags = {
    format: flags.boolean({
      description: dedent`
      [default: true]

      Apply formatting to the given file.
      `,
      allowNo: true,
      default: true,
      char: 'F'
    }),
    write: flags.boolean({
      description: dedent`
      [default: false]

      Write the collected resources to disk.

      A filename can be provided using the --file flag.
      A sensible filename based off the collected resources will be used if one isn't provided.
      `,
      allowNo: true,
      default: false,
      char: 'w'
    }),
    filepath: flags.string({
      description: dedent`
      The name & path of the file to write the Terraform configuration to.

      If the file already exists, the configuration will be appended to the end
      of the file. Otherwise, a new file will be created.
      `,
      required: false,
      char: 'f',
      dependsOn: ['write']
    })
  };

  protected abstract _primaryResourceBlockType: DataType;
  protected abstract _defaultFilenamePrefix: string;

  protected abstract _parse(): ParseResults;

  protected abstract async _nadoResource(
    args: Record<string, unknown>
  ): Promise<TFBlock[]>;

  private _getNameOfFirstResourceBlockOfType(
    blocks: TFBlocks,
    type: DataType
  ): string {
    const resource = blocks.find(
      (block): block is TFResourceBlock =>
        block.blockType === 'resource' &&
        block.labels[0].value === type &&
        block.labels.length === 2 // jic
    );

    if (!resource) {
      throw new Error(`Cannot find block of type ${type}`);
    }

    return resource.labels[1].value;
  }

  protected _buildDefaultFilename(contents: TFFileContents): string {
    return `${
      this._defaultFilenamePrefix
    }_${this._getNameOfFirstResourceBlockOfType(
      contents.blocks,
      this._primaryResourceBlockType
    )}.tf`;
  }

  public async run(): Promise<void> {
    const {
      args,
      flags: {
        format: shouldFormat, //
        filepath: preferredFilePath,
        write: shouldWrite
      }
    } = this._parse();

    let contents: TFFileContents = {
      surroundingText: { leadingOuterText: '', trailingOuterText: '' },
      blocks: await this._nadoResource(args)
    };

    if (contents.blocks.length === 0) {
      return;
    }

    if (shouldFormat) {
      contents = format(contents);
    }

    const output = printTFFileContents(contents);

    if (shouldWrite) {
      const filePath =
        preferredFilePath ?? this._buildDefaultFilename(contents);

      await fs.appendFile(filePath, output);

      this.log(`Written resources to ${filePath}`);

      return;
    }

    this.log(output);
  }
}
