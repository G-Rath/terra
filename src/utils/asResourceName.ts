/**
 * Converts the given `str` to a `Resource` name, by normalising it.
 *
 * - Dots are replaced with underscores
 * - `*` & `\052` are replaced with `_wild_`
 * - Strings starting with numbers are prefixed with an underscore
 *
 * @param {string} str
 *
 * @return {string}
 */
export const asResourceName = (str: string): string =>
  str
    .replace(/\./gu, '_')
    .replace(/\*/gu, '_wild_')
    .replace(/\\052/gu, '_wild_')
    .replace(/^(\d)/u, '_n_$1');
