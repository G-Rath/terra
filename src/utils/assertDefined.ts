export function assertDefined(val: unknown, message?: string): asserts val {
  if (val === undefined) {
    throw new Error(message ?? `Unexpected undefined value`);
  }
}
