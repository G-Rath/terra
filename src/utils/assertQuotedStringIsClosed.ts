/**
 * Asserts that the given `str` has the appropriate closing quote,
 * if it starts with an opening string quote.
 */
export const assertQuotedStringIsClosed = (str: string): void => {
  const quote = str[0];

  if (
    (quote === "'" || quote === '"') &&
    (!str.endsWith(quote) || str.length === 1)
  ) {
    throw new Error(`missing closing ${quote} quote: "${str}"`);
  }
};
