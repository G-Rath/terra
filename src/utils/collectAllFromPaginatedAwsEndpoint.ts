/**
 * Collects all the data from the given AWS Client endpoint by iterating
 * over the paginated results until the `tokenProp` is no longer in the response.
 *
 * @param {(request: TRequest) => Request<TResponse, AWSError>} apiCaller
 * @param {keyof TResponse} tokenProp
 * @param {(previousResponse?: TResponse) => TRequest} requester
 * @param {(currentResponse: TResponse) => TType[]} collector
 * @param {TType[]} results
 *
 * @return {Promise<TType[]>}
 *
 * @template TType
 */
export const collectAllFromPaginatedAwsEndpoint = async <
  TType,
  TReq extends object,
  TRes extends object
>(
  apiCaller: (request: TReq) => { promise: () => Promise<TRes> },
  tokenProp: keyof Omit<TRes, '$response'>,
  requester: (previous: TRes | undefined) => TReq,
  collector: (response: TRes) => TType[],
  results: TType[] = []
): Promise<TType[]> => {
  let response: TRes | undefined;

  do {
    // eslint-disable-next-line no-await-in-loop
    response = await apiCaller(requester(response)).promise();

    results.push(...collector(response));
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  } while (response[tokenProp]);

  return results;
};
