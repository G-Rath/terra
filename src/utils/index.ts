export * from './AwsResourceType';
export * from './asResourceName';
export * from './makeTFArgument';
export * from './makeTFBlockLiteral';
export * from './makeTFDynamicBlock';
export * from './makeTFResourceBlock';
export * from './makeTFStringArgument';
export * from './normaliseRoute53Name';
