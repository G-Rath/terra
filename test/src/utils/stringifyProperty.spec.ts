import { PropertyOrNull, stringifyProperty } from '@src/utils';

/*
terraform {
  required_version = ">= 0.12"
  backend "s3" {
    bucket = "prefer-infrastructure-terraform-state"
    key    = "env-production.tfstate"
    region = "ap-southeast-2"
  }
}
 */

describe('stringifyProperty', () => {
  it.each<[PropertyOrNull]>([
    [['type', '"A"']],
    [['ttl', 300]],
    [['subnet_id', 'aws_subnet.my-subnet.id']],
    [['myProperty', null]],
    [
      [
        // block property
        'terraform', // {
        [['required_version', '">= 0.12"']]
        // }
      ]
    ],
    [
      [
        // block property
        'terraform', // {
        [null, ['required_version', '">= 0.12"'], null]
        // }
      ]
    ],
    [
      [
        // array property
        'records', // = [
        [
          // ['key', 'value'],
          'aws_route53_zone.summerfestival_co_nz.name_servers.0',
          'aws_route53_zone.summerfestival_co_nz.name_servers.1',
          'aws_route53_zone.summerfestival_co_nz.name_servers.2',
          'aws_route53_zone.summerfestival_co_nz.name_servers.3'
        ] // ]
      ]
    ],
    [
      [
        // array property
        'records', // = [
        [
          'aws_route53_zone.summerfestival_co_nz.name_servers.0',
          null, // newline
          'aws_route53_zone.summerfestival_co_nz.name_servers.1',
          null, // newline
          'aws_route53_zone.summerfestival_co_nz.name_servers.2',
          'aws_route53_zone.summerfestival_co_nz.name_servers.3'
        ] // ]
      ]
    ]
  ])('stringifies expectedly', (property: PropertyOrNull) => {
    expect(stringifyProperty(property)).toMatchSnapshot();
  });
});
