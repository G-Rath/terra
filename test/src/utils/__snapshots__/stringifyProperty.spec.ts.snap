// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifyProperty stringifies expectedly 1`] = `"type = \\"A\\""`;

exports[`stringifyProperty stringifies expectedly 2`] = `"ttl = 300"`;

exports[`stringifyProperty stringifies expectedly 3`] = `"subnet_id = aws_subnet.my-subnet.id"`;

exports[`stringifyProperty stringifies expectedly 4`] = `"myProperty = null"`;

exports[`stringifyProperty stringifies expectedly 5`] = `
"terraform {
  required_version = \\">= 0.12\\"
}"
`;

exports[`stringifyProperty stringifies expectedly 6`] = `
"terraform {

  required_version = \\">= 0.12\\"

}"
`;

exports[`stringifyProperty stringifies expectedly 7`] = `
"records = [
  aws_route53_zone.summerfestival_co_nz.name_servers.0,
  aws_route53_zone.summerfestival_co_nz.name_servers.1,
  aws_route53_zone.summerfestival_co_nz.name_servers.2,
  aws_route53_zone.summerfestival_co_nz.name_servers.3,
]"
`;

exports[`stringifyProperty stringifies expectedly 8`] = `
"records = [
  aws_route53_zone.summerfestival_co_nz.name_servers.0,
  null,
  aws_route53_zone.summerfestival_co_nz.name_servers.1,
  null,
  aws_route53_zone.summerfestival_co_nz.name_servers.2,
  aws_route53_zone.summerfestival_co_nz.name_servers.3,
]"
`;
